src/core/plot/qgsvectorlayerplotdatagatherer.cpp

```diff
@@ -0,0 +1,185 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "moc_qgsvectorlayerplotdatagatherer.cpp"
+#include "qgsvectorlayerplotdatagatherer.h"
+#include "qgsexpression.h"
+
+
+QgsVectorLayerXyPlotDataGatherer::QgsVectorLayerXyPlotDataGatherer( const QgsFeatureIterator &iterator, const QgsExpressionContext &expressionContext, const QList<QgsVectorLayerXyPlotDataGatherer::XySeriesDetails> &seriesDetails, Qgis::PlotAxisType xAxisType, const QStringList &predefinedCategories )
+  : mIterator( iterator )
+  , mExpressionContext( expressionContext )
+  , mXAxisType( xAxisType )
+  , mSeriesDetails( seriesDetails )
+  , mPredefinedCategories( predefinedCategories )
+{
+}
+
+bool QgsVectorLayerXyPlotDataGatherer::run()
+{
+  QStringList gatheredCategories;
+  QList<QMap<QString, double>> gatheredSeriesCategoriesSum;
+  QList<QgsXyPlotSeries *> gatheredSeries;
+  for ( int i = 0; i < mSeriesDetails.size(); i++ )
+  {
+    gatheredSeries << new QgsXyPlotSeries();
+    gatheredSeriesCategoriesSum << QMap<QString, double>();
+  }
+
+  QMap<QString, QgsExpression> expressions;
+  QgsFeature feature;
+  while ( mIterator.nextFeature( feature ) )
+  {
+    int seriesIndex = 0;
+    for ( const XySeriesDetails &seriesDetails : mSeriesDetails )
+    {
+      mExpressionContext.setFeature( feature );
+
+      if ( !seriesDetails.filterExpression.isEmpty() )
+      {
+        if ( !expressions.contains( seriesDetails.filterExpression ) )
+        {
+          expressions[seriesDetails.filterExpression] = QgsExpression( seriesDetails.filterExpression );
+          expressions[seriesDetails.filterExpression].prepare( &mExpressionContext );
+        }
+        if ( !expressions[seriesDetails.filterExpression].evaluate( &mExpressionContext ).toBool() )
+        {
+          continue;
+        }
+      }
+
+      if ( !expressions.contains( seriesDetails.xExpression ) )
+      {
+        expressions[seriesDetails.xExpression] = QgsExpression( seriesDetails.xExpression );
+        expressions[seriesDetails.xExpression].prepare( &mExpressionContext );
+      }
+      if ( !expressions.contains( seriesDetails.yExpression ) )
+      {
+        expressions[seriesDetails.yExpression] = QgsExpression( seriesDetails.yExpression );
+        expressions[seriesDetails.yExpression].prepare( &mExpressionContext );
+      }
+
+      bool ok = false;
+      const double y = expressions[seriesDetails.yExpression].evaluate( &mExpressionContext ).toDouble( &ok );
```

Comment:
as above, avoid the extra lookup and use the previously retrieved iterator here (also with x expression later)